File: Problems 0800-0899/838. Push Dominoes/solution.py
class Solution:
    def pushDominoes(self, dominoes: str) -> str:
        # 1. if we see -> <-, then in odd cases, we'll have a standing dominoe
        # in middle
        # 2. if we see <- ->, these don't have effect on each other,
        # only if they make -> <- with another arrow in past/future

        # L at start is for case with first "push" domino. If first push
        # domino is L, then we'll have all L's from start to push domino
        # If first push domino is R, it'll all be stationary
        # R at the end is for case where no dominoes were pushed. We need
        # to fill with all '.'
        dominoes = 'L' + dominoes + 'R'
        # Index of last dominoe that was pushed
        lastPush = 0
        res = ""
        for i in range(1, len(dominoes)):
            if dominoes[i] == '.':
                continue
            
            if lastPush:
                res += dominoes[lastPush]
            numBetween = i - lastPush - 1
            # Same direction
            if dominoes[i] == dominoes[lastPush]:
                res += dominoes[i]*numBetween
            # All standing
            elif dominoes[lastPush] == 'L' and dominoes[i] == 'R':
                res += '.'*numBetween
            else:
                res += 'R'*(numBetween//2)
                res += '.'*(numBetween % 2)
                res += 'L'*(numBetween//2)
        
            lastPush = i
        return res

File: Problems 0800-0899/838. Push Dominoes/solution.cpp
class Solution {
    public:
        string pushDominoes(string dominoes) {
            dominoes = 'L' + dominoes + 'R';
            int lastPush = 0;
            string res = "";
            for (int i = 1; i < dominoes.length(); i++) {
                if (dominoes[i] == '.') continue;
    
                if (lastPush > 0) {
                    res += dominoes[lastPush];
                }
    
                int numBetween = i - lastPush - 1;
                if (dominoes[i] == dominoes[lastPush]) {
                    res += string(numBetween, dominoes[i]);
                } else if (dominoes[i] == 'R' && dominoes[lastPush] == 'L') {
                    res += string(numBetween, '.');
                } else {
                    res += string(numBetween/2, 'R');
                    res += string(numBetween % 2, '.');
                    res += string(numBetween/2, 'L');
                }
    
                lastPush = i;
            }
            return res;
        }
    };